### ISSUE TITLE
Open last document behavior: Intended, bug or feature?

### ISSUE THREAD

**Issue by retrue** [koreader/koreader#2547]
* KOReader version: Last nightly versions
* Device: All of them

#### Steps to reproduce
Open a new book you are not reading. Advance some pages. Close the book. If you check the History, this new book will be the first one in the list.
Now, in the file manager, select this new book and purge sdr. If you check the History, the new book will have disappeared from the list.
Now, choose the option "Open last document". It opens that new book that you just purged.
If instead of that, you close KOReader having ticked before the option "Start with last opened file", next time KOReader will start in that new book that you purged too. In both cases it will open the book showing the cover or first page and with the default configuration.

#### Issue
Is this behaviour normal, the expected one or the most desirable?
If you remove a book from History without purging sdr, when you open it through "Open last document" or "Start with last opened file", the book will open for the last open page respecting the configuration saved in sdr. I think that in this case, it is the most desirable situation.
But it is different if you have purged the book too. My suggestion is that before "Open last document" or "Start with last opened file", the system checks if a sdr folder exists for that document. If that sdr doesn't exist, then open instead the first file that appears in History.

**Comment by Frenzie** [comment]
Also see #2477

**Comment by Frenzie** [comment]
I think the current behavior is basically good and expected, except that "Open last document" should ideally be grayed out. Otherwise doesn't it kind of turn into "open some seemingly random previously opened document"?

**Comment by retrue** [comment]
> I think the current behavior is basically good and expected, except that "Open last document" should ideally be grayed out.

I don't agree that graying out would be the ideal solution. That would be misleading. It would imply the last opened book doesn't exist anymore.
I use the function "Open last document" when I start Koreader to continue reading a book. Sometimes I open a second book to check something, and after that I purge that second book and delete it from history. So, next time I use "Open last document" I expect to open the book that I am really reading, not the book that I checked, purged and deleted from history.

> Otherwise doesn't it kind of turn into "open some seemingly random previously opened document"?

It wouldn't be opening a random previously opened document, it would be always opening the first book in history with a sdr folder.

**Comment by Frenzie** [comment]
> So, next time I use "Open last document" I expect to open the book that I am really reading, not the book that I checked, purged and deleted from history.

I disagree with the expectation. You deleted the last-opened document.

> It wouldn't be opening a random previously opened document, it would be always opening the first book in history with a sdr folder.

*Seemingly* random. The fact that program knows doesn't mean you know where it's coming from.

https://github.com/koreader/koreader/blob/ea2de9638b03b38e52c2f8c6fdf187c9db7b22d5/frontend/apps/filemanager/filemanagermenu.lua#L249 over here it could also say "open History" instead of just displaying a dialog.

**Comment by retrue** [comment]
In the example, I didn't delete the last-opened document: I purged it and removed it from history. If I had deleted it, graying out "open last document" could make sense. But because I didn't delete it from the ereader, it was opened with "open last document".
In the case that the last opened book has been deleted from the ereader, then displaying this "open History" could be handy. But that is a different situation that the explained one in this issue.

**Comment by Frenzie** [comment]
Ah, right. I wouldn't use purging as a shorthand for removing .sdr (admittedly it's called "purge .sdr" which is reasonable enough) because purging would be to remove the book plus anything associated with it (such as .sdr). Basically purging means to remove every trace. ;-)

Nevertheless, I'm not sure if it makes sense to make "Open last document" mean anything other than "Open last document". Especially since the main purpose of purging .sdr is probably to easily get rid of messed up settings? (Margins, font, etc.)

**Comment by poire-z** [comment]
Yes, the purpose of "Purge .sdr" was mostly to get rid of these settings, page positions, bookmarks, reseting a book (mostly when just opening to check if they look fine, and keep it for later - or while developing).
And I added the removal from history (which has nothing to do with "purge" and "sdr" i admit) because that was convenient for my usage.
If that usage (remove every trace except the book) is appreciated, we could just add one more thing:
If the "purged .sdr"'ed book is the last_open_file, just replace last_open_file with the one at the top of the history (after this book has been cleaned from history) ?

**Comment by Frenzie** [comment]
That would be how I'd implement it. I'm not entirely convinced it's the right thing, but have at it I suppose considering it's kind of an edge case.
